Add support for using GPS heading on multirotors

fiam opened the pull request:
Uses the same logic as FW GPS heading, but applies a correction to
the heading reported by the GPS by using the tilt angle of the MR
head.

Please, be very careful testing this feature for the moment.

This commit introduces the new setting use_gps_heading, with
the following values:

- AUTO (default): Enabled just for FW
- ON: Enabled for all crafts
- OFF: Disabled for all crafts

fiam committed:
Highly experimental support for using GPS heading on MR

Uses the same logic as FW GPS heading, but applies a correction to
the heading reported by the GPS by using the tilt angle of the MR
head.

Please, be very careful testing this feature for the moment.

This commit introduces the new setting use_gps_heading, with
the following values:

- AUTO (default): Enabled just for FW
- ON: Enabled for all crafts
- OFF: Disabled for all crafts

digitalentity commented on src/main/flight/imu.c:
```diff
@@ -457,41 +457,78 @@ static void imuCalculateEstimatedAttitude(float dT)
+                // For now, we also require a harcoded speed of 3m/s, but we
+                // should adjust this depending on the maximum speed in modes which do limit
+                // it (e.g. RTH).
+                const uint16_t minAttitudeAngle = DECIDEGREES_TO_DEGREES(10);
```
You can check `calculateCosTiltAngle()` which is essentially generalized tilt `cos(roll)*cos(pitch)`

digitalentity commented on src/main/flight/imu.c:
```diff
@@ -457,41 +457,78 @@ static void imuCalculateEstimatedAttitude(float dT)
+            case GPS_HEADING_AUTO:
+                canUseCOG = STATE(FIXED_WING);
+                break;
+            case GPS_HEADING_ENABLED:
```
We need to verify that MAG is preferred if available.

digitalentity reviewed (commented):
Nice!

fiam committed:
Fix missing word in comment

fiam committed:
Use calculateCosTiltAngle() to check the tilt during COG calculations

As suggested by @digitalentity

Co-authored-by: Konstantin Sharlaimov <konstantin.sharlaimov@gmail.com>

fiam committed:
Enable NAV boxes when using GPS heading on MR

We were only enabling them for FW, since the code checked
explicitely for STATE(FIXED_WING).

fiam committed:
Fix isImuHeadingValid() to work property with GPS heading on MR

Check for imuHasGPSHeadingEnabled() rather than for STATE(FIXED_WING)

fiam committed:
Fix Course over Ground rotation calculation

Use deg/10 rather than deg/100
Use an int16_t rather than uint16_t, since the rotation might go
negative.

fiam committed:
Increase minimum tilt angle for using CoG in MR to 20º

Up from 10º

fiam committed:
Make sure each GPS ground course reading is used only once

Using it multiple times would introduce additional error

fiam committed:
Increase minimum ground speed for COG heading in MR to 6 m/s

3 m/s seems to low and makes it too easy to get false readings.
6 ms/s should still be below speed during RTH in most cases.

fiam committed:
Remove unnecessary clipping of groundCourse after correction

imuMahonyAHRSupdate() already translates it to [-PI, PI].

fiam committed:
Restrict COG heading updates in MR using ACC

Require either a small acceleration or the acceleration direction
to be within PI/4 of the tilt direction to accept the heading.

fiam committed:
Avoid double-checking ground speed for COG heading

We already have platform specific checks, so we don't need any
previous checks for the value.

fiam committed:
Relax yaw checks in RTH_HOVER_PRIOR_TO_LAND when using GPS heading

Add imuHasPreciseHeading(), which returns true iff the craft is
using a true compass for heading. Change the check for the yaw
direcion so it checks for a precise heading sensor rather than
checking directly for FIXED_WING.

fiam committed:
Explicitely initialize gpsHasCOG

Since it's in FASTRAM it needs explicit initialization

fiam committed:
Use estimated speed direction to compensate for COG
